libafl/src/executors/forkserver.rs

```diff
@@ -488,27 +488,25 @@ impl Forkserver {
     }
 
     /// Read from the st pipe
-    pub fn read_st(&mut self) -> Result<(usize, i32), Error> {
+    pub fn read_st(&mut self) -> Result<i32, Error> {
         let mut buf: [u8; 4] = [0_u8; 4];
-
-        let rlen = self.st_pipe.read(&mut buf)?;
+        self.st_pipe.read_exact(&mut buf)?;
         let val: i32 = i32::from_ne_bytes(buf);
-        Ok((rlen, val))
+        Ok(val)
     }
 
     /// Read bytes of any length from the st pipe
-    pub fn read_st_size(&mut self, size: usize) -> Result<(usize, Vec<u8>), Error> {
-        let mut buf = vec![0; size];
-
-        let rlen = self.st_pipe.read(&mut buf)?;
-        Ok((rlen, buf))
+    pub fn read_st_size(&mut self, size: usize) -> Result<Vec<u8>, Error> {
+        let mut buf = vec![0u8; size];
+        self.st_pipe.read_exact(&mut buf)?;
+        Ok(buf)
     }
 
     /// Write to the ctl pipe
-    pub fn write_ctl(&mut self, val: i32) -> Result<usize, Error> {
-        let slen = self.ctl_pipe.write(&val.to_ne_bytes())?;
-
-        Ok(slen)
+    pub fn write_ctl(&mut self, val: i32) -> Result<(), Error> {
+        self.ctl_pipe
```

Comment:
For these I'm actually not sure: writing 4 bytes will succeed in all cases, and we save a few of extra checks here - and it's a very critical path...

Reply:
Let's use normal `.write` but check for `length != 4` and return a result if it failed?

Reply:
It's fine with me. I think we should leave a comment here indicating the risk that the underlying API does not grantee to read/write 4 bytes.

Reply:
Perfect, thank you!